File: solutions/go/linked-list/1/linked_list.go
package linkedlist

import "errors"

type Node struct {
    Value	any
    next	*Node
    prev	*Node
}

type List struct {
    head	*Node
    tail	*Node
}

func NewList(elements ...any) *List {
    size := len(elements)
    if size == 0 {
        return &List{}
    }
    var prev *Node
    var node *Node
    var head *Node
    for i, v := range elements {
        if i == 0 {
            prev = &Node{
                Value:	v,
            }
            head = prev
        } else {
            node = &Node{
                Value:	v,
                prev:	prev,
            }
            prev.next = node
            prev = node
        }
    }
    return &List {
        head:	head,
        tail:	prev,
    }
}

func (n *Node) Next() *Node {
    if n == nil {
        return nil
    }
	return n.next
}

func (n *Node) Prev() *Node {
    if n == nil {
        return nil
    }
	return n.prev
}

func (l *List) Unshift(v any) {
    if l == nil {
        *l = List{}
    }
	first := l.head
    n := &Node{
        Value:	v,
        next:	first,
    }
    if first != nil {
    	first.prev = n   
    } else {
        l.tail = n
    }
    l.head = n
    return
}

func (l *List) Push(v any) {
    if l == nil {
        *l = List{}
    }
	last := l.tail
    n := &Node{
        Value:	v,
        prev:	last,
    }
    if last != nil {
    	last.next = n   
    } else {
        l.head = n
    }
    l.tail = n
    return
}

func (l *List) Shift() (v any, e error) {
    if l == nil {
        *l = List{}
    }
	first := l.head
    if first == nil {
        e = errors.New("shift on empty list")
        return
    }
    if first.next != nil {
        first.next.prev = nil
    } else {
        l.tail = nil
    }
    l.head = first.next
    v = first.Value
    return
}

func (l *List) Pop() (v any, e error) {
    if l == nil {
        *l = List{}
    }
	last := l.tail
    if last == nil {
        e = errors.New("pop on empty list")
        return
    }
    if last.prev != nil {
        last.prev.next = nil
    } else {
        l.head = nil
    }
    l.tail = last.prev
    v = last.Value
    return
}

func (l *List) Reverse() {
    if l == nil {
        *l = List{}
    }
    current := l.head
    for current != nil {
        current.prev, current.next, current = current.next, current.prev, current.next
    }
	l.head, l.tail = l.tail, l.head
    return
}

func (l *List) First() *Node {
    if l == nil {
        return nil
    }
	return l.head
}

func (l *List) Last() *Node {
    if l == nil {
        return nil
    }
	return l.tail
}

File: solutions/javascript/linked-list/1/linked-list.js
export class LinkedList {
  #head
  #len

  constructor() {
    this.#len = 0
  }
  
  [Symbol.iterator]() {
    let cursor = this.#head
    return {
      next() {
        const iter = {
          value: cursor,
          done: cursor === undefined
        }
        cursor = cursor?.next
        return iter
      }
    }
  }

  get tail() {
    let last
    for (const cursor of this) {
      last = cursor
    }
    return last
  }
  
  push(newValue) {
    ++this.#len
    const newNode = {
      value: newValue
    }
    if (this.#head === undefined) {
      this.#head = newNode
      return
    }
    const tail = this.tail
    tail.next = newNode
  }

  pop() {
    if (this.#len === 0) return
    --this.#len
    if (this.#len === 0) {
      const value = this.#head.value
      this.#head = undefined
      return value
    }
    let counter = 0
    for (const cursor of this) {
      if (++counter === this.#len) {
        const value = cursor.next.value
        cursor.next = undefined
        return value
      }
    }
  }

  shift() {
    if (this.#len === 0) return
    --this.#len
    const value = this.#head.value
    if (this.#len === 0) {
      this.#head = undefined
    } else {
      this.#head = this.#head.next
    }
    return value
  }

  unshift(newValue) {
    ++this.#len
    const newNode = {
      value: newValue,
      next: this.#head
    }
    this.#head = newNode
  }

  delete(value) {
    let prev
    for (const cursor of this) {
      if (Object.is(cursor.value, value)) {
        --this.#len
        if (prev === undefined) {
          this.#head = cursor.next
          cursor.next = undefined
        } else {
          prev.next = cursor.next
          cursor.next = undefined
        }
        return
      }
      prev = cursor
    }
  }

  count() {
    return this.#len
  }
}
